velox/py/plan_builder/PyPlanBuilder.h

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/core/PlanNode.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/py/type/PyType.h"
+
+namespace facebook::velox::py {
+
+/// Called from when the Python module is loaded/initialized to register all
+/// resources and structures needed by Velox.
+void registerAllResources();
+
+class PyVector;
+
+/// Wrapper class to expose Velox plan nodes to Python.
+class PyPlanNode {
+ public:
+  explicit PyPlanNode(core::PlanNodePtr planNode = nullptr)
+      : planNode_(std::move(planNode)) {}
+
+  core::PlanNodePtr getPlanNode() const {
+    return planNode_;
+  }
+
+  std::string id() const {
+    return std::string{planNode_->id()};
+  }
+
+  std::string serialize() const {
+    return folly::toJson(planNode_->serialize());
+  }
+
+  std::string name() const {
+    if (!planNode_) {
+      return "[nullptr]";
+    }
+    return std::string(planNode_->name());
+  }
+
+  std::string toString(bool detailed = false, bool recursive = false) const {
+    if (!planNode_) {
+      return "[nullptr]";
+    }
+    return planNode_->toString(detailed, recursive);
+  }
+
+ private:
+  core::PlanNodePtr planNode_;
+};
+
+/// Wrapper class for PlanBuilder. It allows us to avoid exposing all details of
+/// the class to users making it easier to use.
+class PyPlanBuilder {
+ public:
+  PyPlanBuilder(
+      const std::shared_ptr<core::PlanNodeIdGenerator>& generator = nullptr);
+
+  PyPlanNode getPlanNode() const {
+    return PyPlanNode(planBuilder_.planNode());
+  }
+
+  // Returns a new builder sharing the plan node id generator, such that the new
+  // builder can safely be used to build other parts/pipelines of the same plan.
+  PyPlanBuilder newBuilder() {
+    return PyPlanBuilder{planNodeIdGenerator_};
+  }
+
+  /// Add table scan node with basic functionality.
```

Comment:
Is this API specific to Hive Connector? If so, please, clarify. How are you thinking about extending this API to work for any connector?

Is there an assumption that data is read from files. How does this work with delta-files?

Reply:
Good point. For now this is Hive connector specific; we can decouple as needed and as we find new use cases. Will add a comment to clarify.

> Is there an assumption that data is read from files. How does this work with delta-files?

This is encapsulated by the splits you add later on - but we need to find a good API to pass these splits. These will come in the next PR which adds supports for runners.